SWLOR.Game.Server/Service/AreaTemplate.cs

```diff
@@ -0,0 +1,284 @@
+
+        public static int TotalCountAppearancePlaceables { get; set; }
+        public static int TotalCountAppearanceCreatures { get; set; }
+
+        public const int PAGESIZE = 50;
+
+        public static int GetPlaceableAppearancePageCount()
+        {
+            return PlaceableAppearanceByPage.Count;
+        }
+
+        public static int GetCreatureAppearancePageCount()
+        {
+            return CreatureAppearanceByPage.Count;
+        }
+
+        [NWNEventHandler("mod_cache")]
+        public static void CacheAppearances()
+        {
+            TotalCountAppearancePlaceables = 0;
+            TotalCountAppearanceCreatures = 0;            
+
+            var rowCount = UtilPlugin.Get2DARowCount("placeables");
+            var pageNumber = 0;
+            var actualRowCount = 0;
+
+            for (int row = 0; row < rowCount; row++)
+            {
+                var entry = Get2DAString("placeables", "Label", row);
+
+                if (entry != "****" && !string.IsNullOrWhiteSpace(entry))
+                {
+                    if (actualRowCount > 0 && actualRowCount % PAGESIZE == 0) pageNumber++;
+
+                    if (!PlaceableAppearanceByPage.ContainsKey(pageNumber))
+                        PlaceableAppearanceByPage[pageNumber] = new Dictionary<int, string>();
+
+                    PlaceableAppearanceByPage[pageNumber].Add(row, entry);
+                    actualRowCount++;                 
+                }
+            }
+            TotalCountAppearancePlaceables = actualRowCount;
+            Console.WriteLine($"Loaded {PlaceableAppearanceByPage.Count} placeable appearance pages ({AreaTemplate.PAGESIZE} per page)  into cache.");
+
+            rowCount = UtilPlugin.Get2DARowCount("appearance");
+            pageNumber = 0;
+            actualRowCount = 0;
+
+            for (int row = 0; row < rowCount; row++)
+            {
+                var entry = Get2DAString("appearance", "Label", row);
+
+                if (entry != "****" && !string.IsNullOrWhiteSpace(entry))
+                {
+                    if (actualRowCount > 0 && actualRowCount % PAGESIZE == 0) pageNumber++;
+
+                    if (!CreatureAppearanceByPage.ContainsKey(pageNumber))
+                        CreatureAppearanceByPage[pageNumber] = new Dictionary<int, string>();
+
+                    CreatureAppearanceByPage[pageNumber].Add(row, entry);
+                    actualRowCount++;
+                }
+            }
+            TotalCountAppearanceCreatures = actualRowCount;
+            Console.WriteLine($"Loaded {CreatureAppearanceByPage.Count} creature appearance pages ({AreaTemplate.PAGESIZE} per page)  into cache.");
+        }
+
+        [NWNEventHandler("mod_exit")]
+        public static void OnExit()
+        {
+            //return;
+
+            // For some reason I can't determine, the below is not actually working.
```

Comment:
Is this comment still accurate? Is this an outstanding issue?

Reply:
I just tested this and reproduced the issue. The problem is the PC object is already gone from the server by the time this event is called. You can retrieve certain information about the PC but not the actual PC itself. More than likely there's a bunch of stuff that relies on the PC in the GUI framework. So you can't rely on that event firing reliably upon a disconnect.

My recommendation would be to simply remove the effect from the object directly if possible. Alternatively you can do it upon login.